### PR TITLE
Fix and tests provenance variables with prov

### PR DESCRIPTION
**Pull request by nwieters**
- Small bugfix in set_provenance.
- Unit tests for provenance

### CODE REVIEW

**Comment by nwieters** [comment]
Unit tests 9 and 13 still failing. 
- Test 9 is if a new config entry is added as a list.
- Test 13 is to test if setting the provenance of a single entry in a list is working. Don't know if this should be possible in general.

File: src/esm_parser/provenance.py [hunk 1]
```diff
@@ -97,23 +97,51 @@ def wrapper_with_provenance_factory(value, provenance=None):
...
+    Create a subclass of type int
+    """
+
+    def __new__(cls, value, *args, **kwargs):
+        return super(WrapperWithProvenance_int, cls).__new__(cls, value)
+
+    def __init__(self, value, provenance=None):
+             self.provenance = provenance
+
+
```
Comment (mandresm) [inline src/esm_parser/provenance.py]: This belongs to another PR for fixing esm_master

File: src/esm_parser/provenance.py [hunk 2]
```diff
@@ -235,12 +263,11 @@ def set_provenance(self, provenance):
...
         """
-
         for key, val in self.items():
             if isinstance(val, dict):
-                self[key] = DictWithProvenance(val, provenance)
-            if isinstance(val, list):
-                self[key] = ListWithProvenance(val, provenance)
+                DictWithProvenance.set_provenance(val, provenance)
+            elif isinstance(val, list):
+                ListWithProvenance.set_provenance(val, provenance)
```
Comment (mandresm) [inline src/esm_parser/provenance.py]: What you are doing here is using a method of a class without having an object instanciated. I don't think that's what you want to do here.

Instead I would suggest that you first instantiate the object and then you set the provenance:
```suggestion
                self[key] = DictWithProvenance(val, None)
                self[key].set_provenance(val, provenance)
            elif isinstance(val, list):
                self[key] = ListWithProvenance(val, None)
                self[key].set_provenance(val, provenance)

```

File: src/esm_parser/provenance.py [hunk 3]
```diff
@@ -333,12 +360,11 @@ def set_provenance(self, provenance):
...
-
-        for c, elem in enummerate(self):
+        for c, elem in enumerate(self):
             if isinstance(elem, dict):
-                self[c] = DictWithProvenance(elem, provenance)
-            if isinstance(elem, list):
-                self[c] = ListWithProvenance(elem, provenance)
+                DictWithProvenancei.set_provenance(elem, provenance)
+            elif isinstance(elem, list):
+                ListWithProvenance.set_provenance(elem, provenance)
```
Comment (mandresm) [inline src/esm_parser/provenance.py]: Same as above

File: tests/test_esm_parser/test_provenance.py [hunk 4]
```diff
@@ -2,143 +2,129 @@
...
+#        },
+#    }
+#}
+#
+#my_provenance = {
+#    "echam": {
+#        "type": {"line": 2, "col": 11, "yaml_file": "myrunscript.yaml", "category": "runscript"},
+#        },
+#}
+#config = provenance.DictWithProvenance(config_dict, my_provenance)
```
Comment (mandresm) [inline tests/test_esm_parser/test_provenance.py]: Please, remove the comments if we are not going to need those lines anymore.

File: tests/test_esm_parser/test_provenance.py [hunk 5]
```diff
@@ -2,143 +2,129 @@
...
+                'path_in_computer': {'line': 6, 'col': 31, 'yaml_file': str(pathlib.Path(f"{os.environ['GITHUB_WORKSPACE']}/tests/test_esm_parser/example2.yaml").resolve()), 'category': 'runscript'}
+            }
+        }
+    },
+    'debug_info': {'loaded_from_file': None}
 }
-config = provenance.DictWithProvenance(config_dict, my_provenance)
+# Test 1:  Checks for a correct provenance entries from example yaml file.
+def test_get_provenance_from_yaml_to_dict():
+    assert config.get_provenance() == check_provenance
```
Comment (mandresm) [inline tests/test_esm_parser/test_provenance.py]: Turn the comment into the function's docstring
```suggestion
def test_get_provenance_from_yaml_to_dict():
    """
    Checks for a correct provenance entries from example yaml file
    """
    assert config.get_provenance() == check_provenance
```

File: tests/test_esm_parser/test_provenance.py [hunk 6]
```diff
@@ -2,143 +2,129 @@
...
+
+
+# Test 2: Checks whether the provenance of an added config entry is None.
```
Comment (mandresm) [inline tests/test_esm_parser/test_provenance.py]: Same as above

File: tests/test_esm_parser/test_provenance.py [hunk 7]
```diff
@@ -2,143 +2,129 @@
...
+def test_get_provenance_of_added_entry():
+    config["fesom"] = True
+    check_provenance["fesom"] = None
+    assert config.get_provenance() == check_provenance
 
 
-# Test 1 (should give you a provenance of None for the key ["fesom"]["asd"])
-def test_get_provenance_1():
+# Test 3: Checks whether the provenance of an added nested config entry is None.
```
Comment (mandresm) [inline tests/test_esm_parser/test_provenance.py]: Same as above

File: tests/test_esm_parser/test_provenance.py [hunk 8]
```diff
@@ -2,143 +2,129 @@
...
-                "greenhouse": {
-                    "kind": {"from_file": None, "type": None},
-                    "path_in_computer": {"from_file": None, "type": None},
-                }
-            },
-        },
-        "fesom": {"asd": None},
-        "computer": None,
-    }
+# Test 4: Checks whether the provenance of an added nested config entry is None.
```
Comment (mandresm) [inline tests/test_esm_parser/test_provenance.py]: Same as above

File: tests/test_esm_parser/test_provenance.py [hunk 9]
```diff
@@ -2,143 +2,129 @@
...
-                "greenhouse": {
-                    "kind": {"from_file": None, "type": None},
-                    "path_in_computer": {"from_file": None, "type": None},
-                }
-            },
-        },
-        "fesom": {"asd": 2, "model": 2},
-        "computer": None,
-    }
+# Test 5: Reset the provenance of an ``echam`` leave.
```
Comment (mandresm) [inline tests/test_esm_parser/test_provenance.py]: Same as above

File: tests/test_esm_parser/test_provenance.py [hunk 10]
```diff
@@ -2,143 +2,129 @@
...
-                    "kind": {"from_file": None, "type": None},
-                    "path_in_computer": {"from_file": None, "type": None},
-                }
-            },
-        },
-        "fesom": {"asd": 2, "model": 2},
-        "computer": None,
-        True: None,
-    }
+# Test 6: Reset the provenance of leaves for an later added branch ``debug_info``.
```
Comment (mandresm) [inline tests/test_esm_parser/test_provenance.py]: Same as above

File: tests/test_esm_parser/test_provenance.py [hunk 11]
```diff
@@ -2,143 +2,129 @@
...
-    check_provenance = {
-        "echam": {
-            "type": "a_string",
-            "files": {"greenhouse": {"kind": "a_string", "path_in_computer": "a_string"}},
-        },
-        "fesom": {"asd": 2, "model": 2},
-        "computer": None,
-        True: None,
-    }
+# Test 7: Reset the provenance of all ``echam`` leaves to "a_string")
```
Comment (mandresm) [inline tests/test_esm_parser/test_provenance.py]: Same as above

File: tests/test_esm_parser/test_provenance.py [hunk 12]
```diff
@@ -2,143 +2,129 @@
...
-            "type": "a_string",
-            "files": {
-                "greenhouse": {"kind": "a_new_string", "path_in_computer": "a_string"}
-            },
-        },
-        "fesom": {"asd": 2, "model": 2},
-        "computer": None,
-        True: None,
-    }
+# Test 8: Rest the provenanve of fesom entry.
```
Comment (mandresm) [inline tests/test_esm_parser/test_provenance.py]: Same as above

File: tests/test_esm_parser/test_provenance.py [hunk 13]
```diff
@@ -2,143 +2,129 @@
...
+## Test 9: Reset the provenance of a list")
+#def test_set_provenance_for_a_list_leaf():
+#    new_prov = {'line': 2, 'col': 11, 'yaml_file': 'someother.yaml', 'category': 'this_is_for_a_list'}
+#    config["fesom"] = provenance.DictWithProvenance(config["fesom"], {})
+#    config["fesom"] = {"list": [30, 19]}
+#    config["fesom"]["list"] = provenance.ListWithProvenance(config["fesom"]["list"], [None, None])
+#    config["fesom"]["list"].set_provenance(new_prov)
+#    #print("provenance: ",config["fesom"]["list"].get_provenance())
+#    #print("provenance2: ",config.get_provenance())
+#    assert config.get_provenance() == check_provenance
```
Comment (mandresm) [inline tests/test_esm_parser/test_provenance.py]: Delete if it's not needed anymore

File: tests/test_esm_parser/test_provenance.py [hunk 14]
```diff
@@ -2,143 +2,129 @@
...
 
-def test_extract_dict():
+
+# Test 10: Test the extraction of config for all allowed variable types.
```
Comment (mandresm) [inline tests/test_esm_parser/test_provenance.py]: into docstrings

File: tests/test_esm_parser/test_provenance.py [hunk 15]
```diff
@@ -2,143 +2,129 @@
...
+def test_extract_dict_config():
     esm_tools_loader = yaml_to_dict.EsmToolsLoader()
-    file_path = pathlib.Path("example.yaml")
+    os.environ['USER'] = "some_user"
+    file_path = pathlib.Path(str(pathlib.Path(f"{os.environ['GITHUB_WORKSPACE']}/tests/test_esm_parser/example.yaml")))
```
Comment (mandresm) [inline tests/test_esm_parser/test_provenance.py]: I think this will break if run locally. What about having a check here whether `os.environ['GITHUB_WORKSPACE']` exists. If it doesn't then you can use the line `file_path = pathlib.Path("example.yaml")`

File: tests/test_esm_parser/test_provenance.py [hunk 16]
```diff
@@ -164,122 +150,172 @@ def test_extract_dict():
...
 
-    assert data2 == provenance
+    config = provenance.DictWithProvenance(data, data2)
+    assert config["person"]["my_other_list"].get_provenance() == check_prov
 
 
-# Test 7 (reset the provenance of a leaf)
-def test_get_provenance_7():
-    config_fesom = provenance.DictWithProvenance({"fesom": {"update_test": True}}, "new_provenance")
+# Test 12: Check set_provenance of a list entry
```
Comment (mandresm) [inline tests/test_esm_parser/test_provenance.py]: into docstrings

File: tests/test_esm_parser/test_provenance.py [hunk 17]
```diff
@@ -164,122 +150,172 @@ def test_extract_dict():
...
-                                    'category': 'None'}, {
-                                    'my_dict': {
-                                        'foo': {
-                                            'line': 27,
-                                            'col': 17,
-                                            'yaml_file': file_path,
-                                            'category': 'None'}
-                                    }}]}}]}}
+    file_path = pathlib.Path(str(pathlib.Path(f"{os.environ['GITHUB_WORKSPACE']}/tests/test_esm_parser/example.yaml")))
+    check_prov = [{'line': 15, 'col': 19, 'yaml_file': str(pathlib.Path(f"{os.environ['GITHUB_WORKSPACE']}/tests/test_esm_parser/example.yaml").resolve()), 'category': 'runscript'}, {'line': 15, 'col': 22, 'yaml_file': str(pathlib.Path(f"{os.environ['GITHUB_WORKSPACE']}/tests/test_esm_parser/example.yaml").resolve()), 'category': 'runscript'}, {'line': 15, 'col': 25, 'yaml_file': str(pathlib.Path(f"{os.environ['GITHUB_WORKSPACE']}/tests/test_esm_parser/example.yaml").resolve()), 'category': 'runscript'}]
```
Comment (mandresm) [inline tests/test_esm_parser/test_provenance.py]: Potentially the same problem for `os.environ['GITHUB_WORKSPACE']` when running the tests locally. It's probably worth it to define a variable outside of the functions that define the path to the file you are trying to reach depending on an `if os.environ['GITHUB_WORKSPACE']`

File: tests/test_esm_parser/test_provenance.py [hunk 18]
```diff
@@ -164,122 +150,172 @@ def test_extract_dict():
...
-        "computer": None,
-        True: None,
-    }
-    assert config.get_provenance() == check_provenance
+    config = provenance.DictWithProvenance(data, data2)
+    config["person"]["my_other_list"].set_provenance(new_prov)
+    assert config["person"]["my_other_list"].get_provenance() == check_prov
+
+
+# Test 13: Check set_provenance of a single list entry
```
Comment (mandresm) [inline tests/test_esm_parser/test_provenance.py]: into docstrings

File: tests/test_esm_parser/test_provenance.py [hunk 19]
```diff
@@ -164,122 +150,172 @@ def test_extract_dict():
...
+#            "type": "a_string",
+#            "files": {
+#                "greenhouse": {"kind": "a_new_string", "path_in_computer": "a_string"}
+#            },
+#        },
+#        "fesom": {"asd": 2, "model": 2, "update_test": "new_provenance"},
+#        "computer": None,
+#        True: None,
+#    }
+#    assert config.get_provenance() == check_provenance
```
Comment (mandresm) [inline tests/test_esm_parser/test_provenance.py]: Delete if not needed

File: tests/test_esm_parser/test_provenance.py [hunk 20]
```diff
@@ -164,122 +150,172 @@ def test_extract_dict():
...
+    config = provenance.DictWithProvenance(data, data2)
+    config["person"]["my_other_list"].set_provenance(new_prov)
+    assert config["person"]["my_other_list"].get_provenance() == check_prov
+
+
+# Test 13: Check set_provenance of a single list entry
+def test_check_set_provenance_of_single_list_entry():
+    os.environ['USER'] = "some_user"
+    esm_tools_loader = yaml_to_dict.EsmToolsLoader()
+    file_path = pathlib.Path(str(pathlib.Path(f"{os.environ['GITHUB_WORKSPACE']}/tests/test_esm_parser/example.yaml")))
```
Comment (mandresm) [inline tests/test_esm_parser/test_provenance.py]: same as above with the locally run tests

**Review by mandresm — CHANGES_REQUESTED** [review]
Thanks a lot @nwieters, looks great. I have some minor suggestions.

I would also recommend that the tests should run both locally and in GitHub.

File: src/esm_parser/provenance.py (hunk 1 above)
Comment (nwieters) [reply src/esm_parser/provenance.py]: I removed this again, since it is not needed anymore (because no pickle file is created anymore with PR #1020).

File: src/esm_parser/provenance.py (hunk 2 above)
Comment (nwieters) [reply src/esm_parser/provenance.py]: Implemented this with small changes.

File: src/esm_parser/provenance.py (hunk 3 above)
Comment (nwieters) [reply src/esm_parser/provenance.py]: Same as above

File: tests/test_esm_parser/test_provenance.py (hunk 4 above)
Comment (nwieters) [reply tests/test_esm_parser/test_provenance.py]: Removed

File: tests/test_esm_parser/test_provenance.py (hunk 5 above)
Comment (nwieters) [reply tests/test_esm_parser/test_provenance.py]: Done

File: tests/test_esm_parser/test_provenance.py (hunk 16 above)
Comment (nwieters) [reply tests/test_esm_parser/test_provenance.py]: Done

File: tests/test_esm_parser/test_provenance.py (hunk 13 above)
Comment (nwieters) [reply tests/test_esm_parser/test_provenance.py]: Done

File: tests/test_esm_parser/test_provenance.py (hunk 15 above)
Comment (nwieters) [reply tests/test_esm_parser/test_provenance.py]: Done

File: tests/test_esm_parser/test_provenance.py (hunk 20 above)
Comment (nwieters) [reply tests/test_esm_parser/test_provenance.py]: Done

**Review by mandresm — APPROVED** [review]

**Comment by mandresm** [comment]
@nwieters, you can go ahead an merge it if you are happy with it. Then I'll jump in tomorrow to merge provenance history and adapt the tests
